From c1c9ea797acf38502b38f472b6e48206e1019b16 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Wed, 5 Sep 2018 16:48:33 +0800
Subject: [PATCH 4548/5242] MLK-18979-2: ASoC: fsl_asrc: add initial check in
 resume

commit  7a97b10d794931353e1e939967b2b2f6a17448c3 from
https://source.codeaurora.org/external/imx/linux-imx.git

If the initialization is not finished, then we input data to
the FIFO will fail, which still cause the error

"output dma task timeout"

So we need to ad initial check in the resume.

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_asrc.c |    9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sound/soc/fsl/fsl_asrc.c b/sound/soc/fsl/fsl_asrc.c
index aacefa9..51c1955 100644
--- a/sound/soc/fsl/fsl_asrc.c
+++ b/sound/soc/fsl/fsl_asrc.c
@@ -1224,6 +1224,8 @@ static int fsl_asrc_runtime_resume(struct device *dev)
 	struct fsl_asrc *asrc_priv = dev_get_drvdata(dev);
 	int i, ret;
 	u32 asrctr;
+	u32 reg;
+	int retry = 10;
 
 	ret = clk_prepare_enable(asrc_priv->mem_clk);
 	if (ret)
@@ -1260,6 +1262,13 @@ static int fsl_asrc_runtime_resume(struct device *dev)
 	regmap_update_bits(asrc_priv->regmap, REG_ASRCTR,
 			   ASRCTR_ASRCEi_ALL_MASK, asrctr);
 
+	/* Wait for status of initialization */
+	do {
+		udelay(5);
+		regmap_read(asrc_priv->regmap, REG_ASRCFG, &reg);
+		reg = (reg >> ASRCFG_INIRQi_SHIFT(0)) & 0x7;
+	} while (!(reg == ((asrctr & 0xE) >> 1)) && --retry);
+
 	return 0;
 
 disable_asrck_clk:
-- 
1.7.9.5

